File: project/src/main/java/euphony/lib/receiver/EuWindows.java
package euphony.lib.receiver;

import java.nio.ByteBuffer;

public class EuWindows {
	public enum Windows {
		RECTANGULAR, TRIANGLAR, HANNING, HAMMING, BLACKMAN, KAISER
	}

	private Windows mWindowType;
	private ByteBuffer mBuffer;
	private int mBufferSize;
	private byte[] mRawData;
	private double KaiserAlph = 32.0;        
	private double KaiserWindowSize = 45;
	private double HammingAlph = 54.0;   
	private double BlackmanAlph = 0.16;

	public EuWindows(Windows type, ByteBuffer buffer, int bufferSize)
	{
		mBuffer	=	buffer;
		mBufferSize	=	bufferSize;
		mRawData = buffer.array();
		mWindowType = type;
	}

	void Processor()
	{
		if ((mBuffer != null)&&(mBufferSize != 0))// check buffer buffersize
		{
			switch (mWindowType)
			{
			case RECTANGULAR:
				RectangularWindow();
				break;
			case TRIANGLAR:
				TrianglarWindow();
				break;	
			case HANNING:
				HanningWindow();
				break;
			case HAMMING:
				HammingWindow();
				break;
			case BLACKMAN:
				BlackmanWindow();
				break;
			case KAISER:
				KaiserWindow();
				break;
			}
		}
	}
	
	private void RectangularWindow()
	{
		for(int i = 0; i < mRawData.length; i++)
			mRawData[i] = (byte) ( mRawData[i] * 1.0 );
		mBuffer = ByteBuffer.wrap(mRawData);
	
	}
	
	private void TrianglarWindow(){
		for(int i = 0; i < mRawData.length; i++)
			mRawData[i] = (byte) ( ( mRawData[i] ) * (( mRawData.length / 2.0 - Math.abs( i - ( mRawData.length - 1.0 ) / 2.0 ) ) / ( mRawData.length / 2.0 ))) ;
		mBuffer = ByteBuffer.wrap(mRawData);
	}
		
	private void HanningWindow()
	{
		for(int i = 0; i < mRawData.length; i++)
			mRawData[i] = (byte)( ( mRawData[i] ) * ( 0.5 * ( 1 - Math.cos( ( 2 * Math.PI * i ) / ( 512 - 1 ) ) ) ) ) ;	
		mBuffer = ByteBuffer.wrap(mRawData);
	}

	private void HammingWindow(){
		for(int i = 0; i < mRawData.length; i++)
			mRawData[i] = (byte)( ( mRawData[i] ) * ( HammingAlph + ( 1.0 - HammingAlph ) * Math.cos( 2.0 * Math.PI * i / mRawData.length + Math.PI ) )) ;
		mBuffer = ByteBuffer.wrap(mRawData);    
	}

	private void BlackmanWindow(){
		for(int i = 0; i < mRawData.length; i++)
			mRawData[i] = (byte) ( ( mRawData[i] ) * (( 1.0 - BlackmanAlph ) / 2.0 - 0.5 * Math.cos( 2.0 * Math.PI * i / ( mRawData.length - 1.0 ) ) 
			+ ( BlackmanAlph / 2.0 ) * Math.cos( 4.0 * Math.PI * i / ( mRawData.length - 1.0 ) ))) ;
		mBuffer = ByteBuffer.wrap(mRawData);
	}
	
	private void KaiserWindow(){
		double n = KaiserWindowSize / 2.0;
		for(int i = 0; i < mRawData.length; i++)
		{
			mRawData[i] = (byte) ( ( mRawData[i] ) * (Util.I0( KaiserAlph * Math.sqrt( 1.0 - Math.pow(2.0 * ( (double) i - n ) / KaiserWindowSize , 2 ) ) ) / Util.I0(KaiserAlph)));		
		}
		mBuffer = ByteBuffer.wrap(mRawData);
	}
	
	static class Util{
		public static double I0(double x)
		{
			double	eps = 10E-9;
			int	n = 1;
			double	S = 1.0;
			double	D = 1.0;
			while (D > eps * S)
			{
				double	T = x / (2.0 * n);
				n++;
				D *= (T * T);
				S += D;
			}
			return S;
		}
	}
	
	public ByteBuffer getBuffer() {
		return mBuffer;
	}

	public void setBuffer(ByteBuffer mBuffer) {
		this.mBuffer = mBuffer;
	}

	public int getBufferSize() {
		return mBufferSize;
	}

	public void setBufferSize(int mBufferSize) {
		this.mBufferSize = mBufferSize;
	}

	public Windows getWindowType() {
		return mWindowType;
	}

	public void setWindowType(Windows type) { mWindowType = type; }
	public double getKaiserAlph() {
		return KaiserAlph;
	}

	public void setKaiserAlph(double kaiserAlph) {
		KaiserAlph = kaiserAlph;
	}

	public double getKaiserWindowSize() {
		return KaiserWindowSize;
	}

	public void setKaiserWindowSize(double kaiserWindowSize) {
		KaiserWindowSize = kaiserWindowSize;
	}

	public double getHammingAlph() {
		return HammingAlph;
	}

	public void setHammingAlph(double hammingAlph) {
		HammingAlph = hammingAlph;
	}	
	
	public double getBlackmanAlph() {
		return BlackmanAlph;
	}

	public void setBlackmanAlph(double blackmanAlph) {
		BlackmanAlph = blackmanAlph;
	}	
}

File: euphony/src/main/kotlin/euphony/lib/receiver/EuWindows.kt
package euphony.lib.receiver

import java.nio.ByteBuffer

class EuWindows(mWindowNumber: Short, buffer: ByteBuffer, bufferSize: Int) {
    var windowNumber: Short = 0

    var buffer: ByteBuffer? = null
    var bufferSize = 0
    private val mRawData: ByteArray
    var kaiserAlph = 32.0
    var kaiserWindowSize = 45.0
    var hammingAlph = 54.0
    var blackmanAlph = 0.16

    init {
        this.buffer = buffer
        this.bufferSize = bufferSize
        mRawData = buffer.array()
        windowNumber = mWindowNumber
    }

    internal fun Processor() {
        if (buffer != null && bufferSize != 0)
        // check buffer buffersize
        {
            when (windowNumber as Int) {
                RECTANGULAR -> RectangularWindow()
                TRIANGLAR -> TrianglarWindow()
                HANNING -> HanningWindow()
                HAMMING -> HammingWindow()
                BLACKMAN -> BlackmanWindow()
                KAISER -> KaiserWindow()
            }
        }
    }

    internal fun RectangularWindow() {
        for (i in mRawData.indices)
            mRawData[i] = (mRawData[i] * 1.0).toByte()
        buffer = ByteBuffer.wrap(mRawData)

    }

    internal fun TrianglarWindow() {
        for (i in mRawData.indices)
            mRawData[i] = (mRawData[i] * ((mRawData.size / 2.0 - Math.abs(i - (mRawData.size - 1.0) / 2.0)) / (mRawData.size / 2.0))).toByte()
        buffer = ByteBuffer.wrap(mRawData)
    }

    internal fun HanningWindow() {
        for (i in mRawData.indices)
            mRawData[i] = (mRawData[i] * (0.5 * (1 - Math.cos(2.0 * Math.PI * i.toDouble() / (512 - 1))))).toByte()
        buffer = ByteBuffer.wrap(mRawData)
    }

    internal fun HammingWindow() {
        for (i in mRawData.indices)
            mRawData[i] = (mRawData[i] * (hammingAlph + (1.0 - hammingAlph) * Math.cos(2.0 * Math.PI * i.toDouble() / mRawData.size + Math.PI))).toByte()
        buffer = ByteBuffer.wrap(mRawData)
    }

    internal fun BlackmanWindow() {
        for (i in mRawData.indices)
            mRawData[i] = (mRawData[i] * ((1.0 - blackmanAlph) / 2.0 - 0.5 * Math.cos(2.0 * Math.PI * i.toDouble() / (mRawData.size - 1.0)) + blackmanAlph / 2.0 * Math.cos(4.0 * Math.PI * i.toDouble() / (mRawData.size - 1.0)))).toByte()
        buffer = ByteBuffer.wrap(mRawData)
    }

    internal fun KaiserWindow() {
        val n = kaiserWindowSize / 2.0
        for (i in mRawData.indices) {
            mRawData[i] = (mRawData[i] * (Util.I0(kaiserAlph * Math.sqrt(1.0 - Math.pow(2.0 * (i.toDouble() - n) / kaiserWindowSize, 2.0))) / Util.I0(kaiserAlph))).toByte()
        }
        buffer = ByteBuffer.wrap(mRawData)
    }

    internal object Util {
        fun I0(x: Double): Double {
            val eps = 10E-9
            var n = 1
            var S = 1.0
            var D = 1.0
            while (D > eps * S) {
                val T = x / (2.0 * n)
                n++
                D *= T * T
                S += D
            }
            return S
        }
    }

    companion object {

        //windows
        val RECTANGULAR = 0
        val TRIANGLAR = 1
        val HANNING = 2
        val HAMMING = 3
        val BLACKMAN = 4
        val KAISER = 5
    }
}